libs/vkd3d/command.c

```diff
@@ -6557,69 +6653,113 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetPredication(d3d12_command_li
-
-        if (aligned_buffer_offset & (sizeof(uint64_t) - 1))
-        {
-            WARN("Unaligned predicate argument buffer offset %#"PRIx64".\n", aligned_buffer_offset);
+        if (!d3d12_command_allocator_allocate_scratch_memory(list->allocator,
+                sizeof(uint32_t), sizeof(uint32_t), &scratch))
             return;
-        }
 
-        if (!d3d12_resource_is_buffer(resource))
+        begin_info.sType = VK_STRUCTURE_TYPE_CONDITIONAL_RENDERING_BEGIN_INFO_EXT;
+        begin_info.pNext = NULL;
+        begin_info.buffer = scratch.buffer;
+        begin_info.offset = scratch.offset;
+        begin_info.flags = 0;
+
+        if (list->device->device_info.buffer_device_address_features.bufferDeviceAddress)
         {
-            WARN("Predicate arguments must be stored in a buffer resource.\n");
-            return;
+            /* Resolve 64-bit predicate into a 32-bit location so that this works with
+             * VK_EXT_conditional_rendering. We'll handle the predicate operation here
+             * so setting VK_CONDITIONAL_RENDERING_INVERTED_BIT_EXT is not necessary. */
+            d3d12_command_list_invalidate_current_pipeline(list, true);
+            d3d12_command_list_invalidate_root_parameters(list, VK_PIPELINE_BIND_POINT_COMPUTE, true);
+
+            resolve_args.src_va = d3d12_resource_get_va(resource, aligned_buffer_offset);
+            resolve_args.dst_va = scratch.va;
+            resolve_args.invert = operation != D3D12_PREDICATION_OP_EQUAL_ZERO;
+
+            VK_CALL(vkCmdBindPipeline(list->vk_command_buffer, VK_PIPELINE_BIND_POINT_COMPUTE,
+                    predicate_ops->vk_resolve_pipeline));
+            VK_CALL(vkCmdPushConstants(list->vk_command_buffer, predicate_ops->vk_resolve_pipeline_layout,
+                    VK_SHADER_STAGE_COMPUTE_BIT, 0, sizeof(resolve_args), &resolve_args));
+            VK_CALL(vkCmdDispatch(list->vk_command_buffer, 1, 1, 1));
+
+            src_stages = VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT;
+            src_access = VK_ACCESS_SHADER_WRITE_BIT;
         }
+        else
+        {
+            FIXME_ONCE("64-bit predicates not supported.\n");
 
-        FIXME_ONCE("Predication doesn't support clear and copy commands, "
-                "and predication values are treated as 32-bit values.\n");
+            copy_region.srcOffset = resource->heap_offset + aligned_buffer_offset;
+            copy_region.dstOffset = scratch.offset;
+            copy_region.size = sizeof(uint32_t);
 
-        cond_info.sType = VK_STRUCTURE_TYPE_CONDITIONAL_RENDERING_BEGIN_INFO_EXT;
-        cond_info.pNext = NULL;
-        cond_info.buffer = resource->vk_buffer;
-        cond_info.offset = resource->heap_offset + aligned_buffer_offset;
-        switch (operation)
-        {
-            case D3D12_PREDICATION_OP_EQUAL_ZERO:
-                cond_info.flags = 0;
-                break;
+            VK_CALL(vkCmdCopyBuffer(list->vk_command_buffer,
```

Comment:
Do we need a barrier before the copy?

Reply:
That seems to be taken care of with the resource state barrier.

Reply:
The app should be transitioning the buffer to the predication state beforehand, so we shouldn't need it.